### PR TITLE
Multiple slack buses support

### PR DESCRIPTION
**Pull request by geofjamg**
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*
No

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
Feature

**What is the current behavior?** *(You can also link to an open issue here)*
case_SyntheticUSA.m Matpower grid diverges with OLF (70000 buses component) using a flat start (OK with warm start like in Matpower). After investigation it appears that it is because of a large active power injection at slack bus (because of a large initial imbalance) that cannot be supported a surrounding branches (maximal transmissible power issue).  

```java
Network network = Network.read("case_SyntheticUSA.mat");
LoadFlowParameters parameters = new LoadFlowParameters()
        .setNoGeneratorReactiveLimits(true)
        .setDistributedSlack(false)
        .setVoltageInitMode(LoadFlowParameters.VoltageInitMode.UNIFORM_VALUES);
LoadFlow.run(network, parameters);
```

Output:
```bash
INFO  21:12:30.484 c.p.o.a.o.AcloadFlowEngine |  : Ac loadflow complete on network {CC0 SC0} (result=AcLoadFlowResult(outerLoopIterations=1, newtonRaphsonIterations=31, newtonRaphsonStatus=MAX_ITERATION_REACHED, slackBusActivePowerMismatch=8.667905657473198E11, distributedActivePower=0.0))
```

**What is the new behavior (if this is a feature change)?**

Multiple slack buses are now supported. Reference bus (voltage angle 0°) is taken at first slack bus. All slack buses have an unknown active power injection. Some additional equations are added to define active power distribution between slack buses (proportional). The aim is to spread active power imbalance between multiple slack buses to avoid previously described phenomena (large active power injection not physically transmissible by surrounding branches).

Note: THIS FEATURE HAS NOTHING TO DO WITH DISTRIBUTED SLACK. Slack distribution still work well with multiple slack buses.

So now we have a working code example:

```java
Network network = Network.read("case_SyntheticUSA.mat");
LoadFlowParameters parameters = new LoadFlowParameters()
        .setReadSlackBus(false)
        .setNoGeneratorReactiveLimits(true)
        .setDistributedSlack(false)
        .setVoltageInitMode(LoadFlowParameters.VoltageInitMode.UNIFORM_VALUES);
OpenLoadFlowParameters.create(parameters)
        .setMaxSlackBusCount(3);
LoadFlow.run(network, parameters);
```

Output:
```bash
INFO  21:18:54.029 c.p.o.n.LfNetwork    |  : Network {CC0 SC0}, slack buses are [VL-29184_1, VL-29787_1, VL-43184_1] (method='Most meshed bus')
...
INFO  21:18:56.039 c.p.o.a.o.AcloadFlowEngine |  : Ac loadflow complete on network {CC0 SC0} (result=AcLoadFlowResult(outerLoopIterations=1, newtonRaphsonIterations=5, newtonRaphsonStatus=CONVERGED, slackBusActivePowerMismatch=660.4790932750097, distributedActivePower=0.0))
```

**Does this PR introduce a breaking change or deprecate an API?** *If yes, check the following:*
- [ ] The *Breaking Change* or *Deprecated* label has been added
- [ ] The migration guide has been updated in the github wiki *(What changes might users need to make in their application due to this PR?)*

**Other information**:

(if any of the questions/checkboxes don't apply, please delete them entirely)

### CODE REVIEW

File: src/main/java/com/powsybl/openloadflow/OpenLoadFlowParameters.java [hunk 1]
```diff
@@ -113,6 +113,8 @@ public class OpenLoadFlowParameters extends AbstractExtension<LoadFlowParameters
 
     public static final String STATE_VECTOR_SCALING_MODE_NAME = "stateVectorScalingMode";
 
+    public static final String MAX_SLACK_BUS_COUNT_NAME = "maxSlackBusCount";
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/OpenLoadFlowParameters.java:116]: Why it is a max and not the number of slack buses?

File: src/main/java/com/powsybl/openloadflow/network/FirstSlackBusSelector.java [hunk 2]
```diff
@@ -7,14 +7,15 @@
...
 /**
  * @author Geoffroy Jamgotchian <geoffroy.jamgotchian at rte-france.com>
  */
 public class FirstSlackBusSelector implements SlackBusSelector {
 
     @Override
-    public SelectedSlackBus select(List<LfBus> buses) {
-        return new SelectedSlackBus(buses.get(0), "First bus");
+    public SelectedSlackBus select(List<LfBus> buses, int limit) {
+        return new SelectedSlackBus(buses.stream().limit(limit).collect(Collectors.toList()), "First bus");
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/network/FirstSlackBusSelector.java:19]: Maybe "first bus" is no more appropriate.

File: src/main/java/com/powsybl/openloadflow/network/LargestGeneratorSlackBusSelector.java [hunk 3]
```diff
@@ -32,11 +32,13 @@ private boolean isGeneratorInvalid(LfGenerator generator) {
...
+        List<LfBus> slackBuses = buses.stream()
                 .filter(bus -> !bus.getGenerators().isEmpty() && bus.getGenerators().stream().noneMatch(this::isGeneratorInvalid))
-                .max(Comparator.comparingDouble(LargestGeneratorSlackBusSelector::getMaxP))
-                .orElseThrow(() -> new PowsyblException("Cannot find a bus with a generator"));
-        return new SelectedSlackBus(slackBus, "Largest generator bus");
+                .sorted(Comparator.comparingDouble(LargestGeneratorSlackBusSelector::getMaxP).reversed())
+                .limit(limit)
+                .collect(Collectors.toList());
+
+        return new SelectedSlackBus(slackBuses, "Largest generator bus");
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/network/LargestGeneratorSlackBusSelector.java:42]: Same remark here, bus should becomes plural maybe.

File: src/main/java/com/powsybl/openloadflow/OpenLoadFlowProvider.java [hunk 4]
```diff
@@ -151,7 +151,7 @@ private LoadFlowResult runAc(Network network, LoadFlowParameters parameters, Rep
                                                                             result.getNetwork().getNumSC(),
                                                                             status,
                                                                             result.getNewtonRaphsonIterations(),
-                                                                            slackBusId,
+                                                                            slackBusId, // FIXME manage multiple slack buses
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/OpenLoadFlowProvider.java:154]: Are you going to fix this in that PR?

File: src/main/java/com/powsybl/openloadflow/network/FirstSlackBusSelector.java (hunk 2 above)
Comment (geofjamg) [reply src/main/java/com/powsybl/openloadflow/network/FirstSlackBusSelector.java:19]: I think it is ok to let it singular as it is a name of a method.

File: src/main/java/com/powsybl/openloadflow/network/LfNetwork.java [hunk 5]
```diff
@@ -108,15 +111,24 @@ public void setReporter(Reporter reporter) {
     }
 
     private void invalidateSlack() {
-        slackBus = null;
+        if (slackBuses != null) {
+            for (var slackBus : slackBuses) {
+                slackBus.setSlack(false);
+            }
+            slackBuses.get(0).setReference(false);
```
[the remark this answers is not in the document]
Comment (geofjamg) [reply src/main/java/com/powsybl/openloadflow/network/LfNetwork.java:118]: Same for updateSlackBuses, which also set the reference bus. So we should rename it to invalidateSlackAndReference?

File: src/main/java/com/powsybl/openloadflow/OpenLoadFlowProvider.java (hunk 4 above)
Comment (geofjamg) [reply src/main/java/com/powsybl/openloadflow/OpenLoadFlowProvider.java:154]: That would require to make some change in the loadflow API so not in this PR.

File: src/main/java/com/powsybl/openloadflow/network/LfNetwork.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (annetill) [reply src/main/java/com/powsybl/openloadflow/network/LfNetwork.java:118]: Maybe yes, but not so important indeed.

**Review by annetill — APPROVED** [review]

**Review by annetill — APPROVED** [review]

**Pull request merged**